internal/strquote/strquote.go

```diff
@@ -33,11 +28,13 @@ func Append(buf []byte, s []byte) []byte {
 		case '\\':
 			buf = append(buf, '\\', '\\')
 		default:
-			buf = append(buf, '\\', 'x', hexDigit(b/16), hexDigit(b%16))
+			if needsEscape(b) {
```

Comment:
Thank you for submitting a pull request! I'd much rather see needsEscape() altered to detect backslash, double quote and single quote, preferring to keep both the switch and needsEscape() in sync.

Reply:
Thank you for the review, @davidhubbard. I returned to the previous version of the code and just updated _needsEscape_ to handle the missing characters. I don't find it super readable but that's probably just a preference. Let me know if this is what you imagined.

As a side note, I don't know what kind of a quoted string is this expected to produce. This will produce a string literal that is [valid in Go](https://go.dev/ref/spec#String_literals). However I looked at [quoted string parsing in capnp](https://github.com/capnproto/capnproto/blob/d6f32b38362749be7fdb7edf28b801ba92d56e7a/c%2B%2B/src/capnp/compat/json.c%2B%2B#L777-L785) and the quoting there is incompatible. For instance, `\'` is not a valid escape sequence in capnp.

I don't mind too much, I just want to get the backslash fixed because that is a difference from [zombiezen.com/go/capnproto2](https://pkg.go.dev/zombiezen.com/go/capnproto2).

Reply:
Thanks! Yes, this is what I was hoping for. I understand what you mean about reading the code - it makes sense to me, but I'm still happy for the detailed feedback.

For your question about [quoted string parsing in capnp](https://github.com/capnproto/capnproto/blob/d6f32b38362749be7fdb7edf28b801ba92d56e7a/c%2B%2B/src/capnp/compat/json.c%2B%2B#L777-L785) I think what @lthibault said on [matrix](https://matrix.to/#/!pLcnVUHHRZrUPscloW:matrix.org?via=matrix.org) sums it up: "Do you guys have any thoughts about #546? My understanding of that feature was that it was intended to print data to the screen (a la `Println`), not to *serialize*."

In short, I believe everyone is in favor of this change because:

* It fixes backslash quoting (a regression from https://pkg.go.dev/zombiezen.com/go/capnproto2)
* We're always grateful for pull requests, and this one is well written
* It fixes the underlying issue that `needsEscape()` was not in sync with `Append()`

And I note the caveats we've unconvered:

* This wasn't intended to be fed into capnp string parsing, I think it was just for printing to the screen. Specifically `\'` is not a valid escape sequence in capnp.
* This code could use some love along the lines of becoming more readable (or maybe just some comments?)

Reply:
Thank you for detailed comments and merging.